core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/conditional/CaseExpression.kt

```diff
@@ -0,0 +1,159 @@
+    SimpleElseCaseExpression(case, firstSearchResult, *additionalSearchResult, elseCase = elseCase)
+
+class SearchedCaseExpression<U : ValidType>(
+    val case: SearchedCaseClass,
+    val firstSearchResult: SearchResult<BooleanType, out U>,
+    vararg val additionalSearchResult: SearchResult<BooleanType, out U>,
+) : TypeExpression<U> {
+    override fun toDopeQuery(manager: DopeQueryManager): DopeQuery {
+        val caseDopeQuery = case.toDopeQuery(manager)
+        val whenThenDopeQueries = listOf(firstSearchResult, *additionalSearchResult).map {
+            it.searchExpression.toDopeQuery(manager) to it.resultExpression.toDopeQuery(manager)
+        }
+        return DopeQuery(
+            queryString = caseDopeQuery.queryString +
+                whenThenDopeQueries.joinToString(" ", " ", " ") { "$WHEN ${it.first.queryString} $THEN ${it.second.queryString}" } +
+                END,
+            parameters = whenThenDopeQueries.fold(emptyMap()) { parameters, query ->
+                parameters + query.first.parameters + query.second.parameters
+            },
+        )
+    }
+}
+
+class SearchedCaseClass : Resolvable {
+    override fun toDopeQuery(manager: DopeQueryManager) = DopeQuery(queryString = CASE, parameters = emptyMap())
+}
+
+class SearchedElseCaseExpression<U : ValidType>(
+    private val case: SearchedCaseClass,
+    private val firstSearchResult: SearchResult<BooleanType, out U>,
+    private vararg val additionalSearchResult: SearchResult<BooleanType, out U>,
+    private val elseCase: UnaliasedExpression<out U>,
+) : TypeExpression<U> {
+    override fun toDopeQuery(manager: DopeQueryManager): DopeQuery {
+        val caseDopeQuery = case.toDopeQuery(manager)
+        val whenThenDopeQueries = listOf(firstSearchResult, *additionalSearchResult).map {
+            it.searchExpression.toDopeQuery(manager) to it.resultExpression.toDopeQuery(manager)
+        }
+        val elseCaseDopeQuery = elseCase.toDopeQuery(manager)
+        return DopeQuery(
+            queryString = caseDopeQuery.queryString +
+                whenThenDopeQueries.joinToString(" ", " ", " ") { "$WHEN ${it.first.queryString} $THEN ${it.second.queryString}" } +
+                "$ELSE ${elseCaseDopeQuery.queryString} " +
+                END,
+            parameters = whenThenDopeQueries.fold(emptyMap<String, Any>()) { parameters, query ->
+                parameters + query.first.parameters + query.second.parameters
+            } + elseCaseDopeQuery.parameters,
+        )
+    }
+}
+
+fun case() = SearchedCaseClass()
+
+fun <U : ValidType> SearchedCaseClass.whenThen(whenThenCondition: SearchResult<BooleanType, U>) =
+    SearchedCaseExpression(this, whenThenCondition)
+
+@JvmName("whenThenWithGeneric")
+fun <U : ValidType> SearchedCaseExpression<U>.whenThen(whenThenCondition: SearchResult<BooleanType, U>) =
+    SearchedCaseExpression(case, firstSearchResult, *additionalSearchResult, whenThenCondition)
+
+@JvmName("whenThenWithoutGeneric")
```

Comment:
_whenThenWithoutGeneric_? There is a lot of generic in this function

Reply:
done